// libs/lr2ir/lib/lr2ir.js
"use strict";

const { user, setUser, getUserCensored } = require("./user");
const { isExpired, requestAndConfirm } = require("./request-confirm");
const { LOGIN, ADD_RIVAL, DELETE_RIVAL } = require("./const/service");
const { MY_PAGE, MY_LIST, SONG_RANKING } = require("./const/view");
const prefs = require("./prefs");

const rc = requestAndConfirm;

function Lr2ir() {

	this.setUser = function(lr2id, password) {
		setUser(lr2id, password);
		return this;
	};
	this.getUserCensored = getUserCensored;
	this.prefs = prefs;

	this.login = function(lr2id = user.lr2id, password = user.password) {
		return rc[LOGIN](lr2id, password);
	};

	this.addRival = function(playerid) {
		const loginPromise = isExpired() ? this.login() : Promise.resolve();
		return loginPromise
			.then(() => rc[ADD_RIVAL](playerid))
			.catch((err) => Promise.reject(err));
	};

	this.deleteRival = function(playerid) {
		const loginPromise = isExpired() ? this.login() : Promise.resolve();
		return loginPromise
			.then(() => rc[DELETE_RIVAL](playerid))
			.catch((err) => Promise.reject(err));
	};

	this.open = rc; // TODO
}

const lr2ir = new Lr2ir();

module.exports = lr2ir;
