// app/lr2ir/lr2ir.js
"use strict";

const { user, setUser, getUserCensored } = require("./user");
const { isExpired } = require("./r");
const { request, confirm, requestAndConfirm } = require("./request-confirm");
const SERVICE = require("./const/service");
const VIEW = require("./const/view");
const settings = require("./settings");

const { LOGIN, ADD_RIVAL, DELETE_RIVAL } = SERVICE;
const { MY_PAGE, MY_LIST, SONG_RANKING } = VIEW;
const rc = requestAndConfirm;


function FoonLr2ir() {

	this.setUser = function(lr2id, password) {
		setUser(lr2id, password);
		return this;
	};
	this.getUserCensored = getUserCensored;

	this.raw = {
		request: request,
		confirm: confirm,
		requestAndConfirm: requestAndConfirm
	};

	this.login = function(lr2id = user.lr2id, password = user.password) {
		return rc[LOGIN](lr2id, password);
	};

	this.addRival = function(playerid) {
		const loginPromise = isExpired() ? this.login() : Promise.resolve();
		return loginPromise
			.then(() => rc[ADD_RIVAL](playerid))
			.catch((err) => Promise.reject(err));
	};

	this.deleteRival = function(playerid) {
		const loginPromise = isExpired() ? this.login() : Promise.resolve();
		return loginPromise
			.then(() => rc[DELETE_RIVAL](playerid))
			.catch((err) => Promise.reject(err));
	};

	this.openMyPage = (playerid) => rc[MY_PAGE](playerid);
	this.openMyList = (playerid, sort, page) => rc[MY_LIST](playerid, sort, page);
	this.openSongRanking = (bmsid, page) => rc[SONG_RANKING](bmsid, page);

	// Just another to write. It's the same as openXxXxxx().
	this.open = {};
	Object.values(VIEW).forEach((type) => {
		this.open[type] = rc[type];
	});

	this.settings = settings;
}

const lr2ir = new FoonLr2ir();

module.exports = lr2ir;
