// app/lr2ir-crawler/model/entry.js
"use strict";

/**
 * Represents a record of the ○○のプレイデータ table
 *
 * @class      PlayDataEntry playDataEntry
 * @param      {string}  no       1 2 3 ... 100
 * @param      {string}  title    タイトル
 * @param      {string}  bmsid    bmsid
 * @param      {string}  clear    クリア
 * @param      {string}  ranking  ランキング
 */
function PlayDataEntry(no, title, bmsid, clear, ranking) {
	this.no = no;
	this.title = title;
	this.bmsid = bmsid;
	this.clear = clear;
	// this.playCount = playCount;
	this.ranking = ranking;
}

PlayDataEntry.prototype.toString = function() {
	return `${this.no.padStart(3)} ${this.title} [${this.bmsid}] ${this.clear} ${this.ranking}`;
};


/**
 * Represents a record of the ランキング table
 *
 * @class      RankingEntry rankingEntry
 * @param      {string}  player    プレイヤー
 * @param      {string}  playerid  playerid
 * @param      {string}  score     スコア
 */
function RankingEntry(player, playerid, score) {
	// this.ranking = ranking;
	this.player = player;
	this.playerid = playerid;
	// this.dan = dan;
	// this.clear = clear;
	// this.rank = rank;
	this.score = score;
	// this.combo = combo;
	// ...
	// this.comment = comment;
}

RankingEntry.prototype.toString = function() {
	return `${this.player} [${this.playerid}] ${this.score}`;
};


// Some functions

function isFC(clear) {
	return clear.indexOf("FULLCOMBO") !== -1;
}

function isTop(ranking) {
	return ranking.indexOf("1/") === 0; // Am I missing something?
}

function isMax(score) {
	return score.indexOf("(100%)") !== -1;
}

function comparePlayerID(id1, id2) {
	return id1 === id2; // TODO: 有要支援 1 和 000001 的差別嗎?
}


// Add methods

function addMethods(Fn) {
	const obj = new Fn();
	if (obj.hasOwnProperty("clear"))    Fn.prototype.isFC = () => isFC(this.clear);
	if (obj.hasOwnProperty("ranking"))  Fn.prototype.isTop = () => isTop(this.ranking);
	if (obj.hasOwnProperty("score"))    Fn.prototype.isMax = () => isMax(this.score);
	if (obj.hasOwnProperty("playerid")) Fn.prototype.comparePlayerID = (id) => comparePlayerID(this.playerid, id);
}

addMethods(PlayDataEntry);
addMethods(RankingEntry);


// Add things related to $

const querystring = require("querystring");

PlayDataEntry.createFrom$Row = function($row) {
	const $cells = $row.find("td");
	const $noCell = $cells.eq(0),
	      $titleCell = $cells.eq(1),
	      $clearCell = $cells.eq(2),
	      $rankingCell = $cells.eq(4);
	const no = $noCell.text(),
	      title = $titleCell.text(),
	      href = $titleCell.find("a").first().prop("href"),
	      bmsid = querystring.parse(href).bmsid,
	      clear = $clearCell.text(),
	      ranking = $rankingCell.text();
	const entry = new PlayDataEntry(no, title, bmsid, clear, ranking);
	return entry;
};

PlayDataEntry.createFrom$Table = function($, $table) {
	const $rows = $table.find("tr")
		                .not(":has('th')"); // no header row
	const entries = [];
	$rows.each((i, el) => {
		const $row = $(el);
		const entry = PlayDataEntry.createFrom$Row($row);
		entries.push(entry);
	});
	return entries;
};

RankingEntry.createFrom$Row = function($dataRow /*, $commentRow*/) {
	const $cells = $dataRow.find("td");
	const $playerCell = $cells.eq(1),
	      $scoreCell = $cells.eq(5);
	const player = $playerCell.text(),
	      href = $playerCell.find("a").first().prop("href"),
	      playerid = querystring.parse(href).playerid,
	      score = $scoreCell.text();
	const entry = new RankingEntry(player, playerid, score);
	return entry;
};

RankingEntry.createFrom$Table = function($, $table) {
	const $rows = $table.find("tr")
		                .not(":has('th')") // no header row
		                .not(":has('[colspan]')"); // no comment rows
	const entries = [];
	$rows.each((i, el) => {
		const $row = $(el);
		const entry = RankingEntry.createFrom$Row($row);
		entries.push(entry);
	});
	return entries;
};


// We're done

module.exports = {
	PlayDataEntry,
	RankingEntry
};
