// app/heroku-waker/index.js
"use strict";

const http = require("http");

let url = ""; // Error: Protocol "https:" not supported. Expected "http:"
let interval = 29 * 60 * 1000;

const debug = { // TODO dry
	log: console.log,
	warn: console.log,
	setLogger: function(logger) {
		this.log = logger;
		this.warn = logger;
	},
};

const wakeUp = function(callback) {
	http.get(url, callback);
};
let intervalID = null;

module.exports = {
	setUrl: function(newUrl) {
		url = newUrl;
		return this;
	},
	setInterval: function(newInterval) {
		interval = newInterval;
		return this;
	},
	setLogger: function(newLogger) {
		debug.setLogger(newLogger);
		return this;
	},
	// TODO rename?
	start: function() {
		if (intervalID !== null) {
			debug.warn("heroku-waker: start... already started. ignore.");
			return;
		}
		debug.log("heroku-waker: start.");
		// TODO: setInterval -> setTimout
		intervalID = setInterval(wakeUp, interval);
	},
	stop: function() {
		debug.log("heroku-waker: stop.");
		clearInterval(intervalID);
		intervalID = null;
	},
	wake: wakeUp,
};

/*
Memo

Similar:
https://www.npmjs.com/package/heroku-alive
https://www.npmjs.com/package/hubot-heroku-keepalive
https://github.com/Neamar/heroku-self-ping

https://community.wanikani.com/t/whats-the-difference-between-a-period-and-an-interval/21717/5
	Period:
	A portion of time determined by some recurring phenomena
	The interval of time required for a cyclic motion or phenomenon to complete a cycle and begin to repeat itself
	A chronological division: stage

	Interval:
	A space of time between events
	A space between objects, units or states.
*/
