// app/lr2ir-crawler/util.js
"use strict";

const { 
	PlayDataEntry, 
	RankingEntry 
} = require("./model/entry");

class MyList {
	static getPlayDataTable($) {
		return $("table").first();
	}

	static getPlayDataEntries($) {
		const $table = this.getPlayDataTable($);
		const $rows = $table.find("tr")
		                    .not(":has('th')"); // no header row
		const entries = [];
		$rows.each((i, el) => {
			const $row = $(el);
			const entry = PlayDataEntry.createFromRow($row);
			entries.push(entry);
		});
		return entries;
	}
}

class SongRanking {
	static getRankingTable($) {
		return $("table").eq(3); // 0 情報 1-2 総合ステータス 3 ランキング
	}

	static getRankingEntries($) {
		const $table = this.getRankingTable($);
		const $rows = $table.find("tr")
		                    .not(":has('th')")
		                    .not(":has('[colspan]')"); // no comment rows
		const entries = [];
		$rows.each((i, el) => {
			const $row = $(el);
			const entry = RankingEntry.createFromRow($row);
			entries.push(entry);
		});
		return entries;
	}
}

module.exports = {
	MyList,
	SongRanking
};
